execution_engine_testing/tests/src/test/regression/gov_42.rs

```diff
@@ -0,0 +1,211 @@
+                    .with_session_bytes(input_wasm_bytes.to_vec(), session_args)
+                    .with_empty_payment_bytes(runtime_args! {ARG_AMOUNT => payment_amount,}),
+                expected_error,
+            ),
+        };
+        let deploy = deploy_item_builder
+            .with_address(account_hash)
+            .with_authorization_keys(&[account_hash])
+            .with_deploy_hash(deploy_hash)
+            .build();
+
+        (
+            ExecuteRequestBuilder::new().push_deploy(deploy),
+            expected_error_message,
+        )
+    };
+    let do_minimum_request = do_minimum_request_builder.build();
+
+    let mut builder = InMemoryWasmTestBuilder::default();
+    builder.run_genesis(&PRODUCTION_RUN_GENESIS_REQUEST);
+
+    let account = builder.get_account(*DEFAULT_ACCOUNT_ADDR).unwrap();
+
+    let proposer_balance_before = builder.get_proposer_purse_balance();
+
+    let account_balance_before = builder.get_purse_balance(account.main_purse());
+
+    let empty_wasm_in_payment = match execution_phase {
+        ExecutionPhase::Payment => input_wasm_bytes.is_empty(),
+        ExecutionPhase::Session => false,
+    };
+
+    if empty_wasm_in_payment {
+        // Special case: We expect success, since default payment will be used instead.
+        builder.exec(do_minimum_request).expect_success().commit();
+    } else {
+        builder.exec(do_minimum_request).expect_failure().commit();
+
+        let actual_error = builder.get_error().expect("should have error").to_string();
+        assert!(actual_error.contains(expected_error_message));
+
+        let gas = builder.last_exec_gas_cost();
+        assert_eq!(gas, Gas::zero());
+
+        let account_balance_after = builder.get_purse_balance(account.main_purse());
+        let proposer_balance_after = builder.get_proposer_purse_balance();
+
+        assert_eq!(account_balance_before - *MAX_PAYMENT, account_balance_after);
+        assert_eq!(
+            proposer_balance_before + *MAX_PAYMENT,
+            proposer_balance_after
+        );
+    }
+}
+
+#[ignore]
+#[test]
+fn should_charge_session_with_incorrect_wasm_files() {
+    // This test focuses on testing whether we charge for
+    // WASM files that are malformed (unparseable).
+
+    // If we're provided with such file, we should charge.
+    // The expection here is the "empty wasm" send as
+    // a payment, because in such case we use the "default payment"
+    // instead.
+
+    // For increased security, we also test some other cases in this test
+    // like gas overflow (which is a runtime error).
+
+    // TODO:
+    // Introduce proper layers in wasm execution, for example:
+    // Layer 1. Parse (may throw errors related to malformed wasm bytes)
+
+    // Layer 2. Validation (may throw errors for parseable, but unsupported wasm files, for example,
+    // without memory section)
+
+    // Layer 3. Execution (may throw errors strictly related to
+    // execution for example, stack overflow)
```

Comment:
I'd rather this weren't a todo list in the test.  Let's discuss and create tickets as required.

Reply:
Comments removed here: 4e18358bba6f4e10e4071536a2ad2d1d262db4c8